tidy3d/components/grid/grid_spec.py

```diff
@@ -529,7 +531,138 @@ def _make_coords_initial(
+
+    def _processing_structures(self, structures: List[StructureType]) -> List[StructureType]:
+        """Processing structure list before passing to ``mesher``. Adjust all structures to have
+        step size ``dl``.
+        """
+        processed_structures = []
+        for structure in structures:
+            dl = [self.dl, self.dl, self.dl]
+            # pay attentition to override structures containing dl = None
+            if isinstance(structure, MeshOverrideStructure):
+                for ind, dl_axis in enumerate(structure.dl):
+                    if dl_axis is None:
+                        dl[ind] = None
+            processed_structures.append(MeshOverrideStructure(geometry=structure.geometry, dl=dl))
+        return processed_structures
+
+    @property
+    def _dl_min(self) -> float:
+        """Lower bound of grid size."""
+        if self.dl_min is None or self.dl_min == 0:
+            return 0.5 * self.dl
+        return self.dl_min
+
+    @property
+    def _min_steps_per_wvl(self) -> float:
+        """Minimal steps per wavelength."""
+        # irrelevant in this class, just supply an arbitrary number
+        return 1
+
+    def _dl_max(self, sim_size: Tuple[float, 3]) -> float:
+        """Upper bound of grid size."""
+        return self.dl
+
+    def _single_grid_step_size(self, wavelength: float) -> float:
+        """The grid step size if just a single grid along an axis."""
+        return self.dl
+
+
+class AutoGrid(AbstractAutoGrid):
+    """Specification for non-uniform grid along a given dimension.
+
+    Example
+    -------
+    >>> grid_1d = AutoGrid(min_steps_per_wvl=16, max_scale=1.4)
+
+    See Also
+    --------
+
+    :class:`UniformGrid`
+        Uniform 1D grid.
+
+    :class:`GridSpec`
+        Collective grid specification for all three dimensions.
+
+    **Notebooks:**
+        * `Using automatic nonuniform meshing <../../notebooks/AutoGrid.html>`_
+
+    **Lectures:**
+        *  `Time step size and CFL condition in FDTD <https://www.flexcompute.com/fdtd101/Lecture-7-Time-step-size-and-CFL-condition-in-FDTD/>`_
+        *  `Numerical dispersion in FDTD <https://www.flexcompute.com/fdtd101/Lecture-8-Numerical-dispersion-in-FDTD/>`_
+    """
+
+    min_steps_per_wvl: float = pd.Field(
+        10.0,
+        title="Minimal Number of Steps Per Wavelength",
+        description="Minimal number of steps per wavelength in each medium.",
+        ge=6.0,
+    )
+
+    min_steps_per_sim_size: float = pd.Field(
+        10.0,
+        title="Minimal Number of Steps Per Simulation Domain Size",
+        description="Minimal number of steps per longest edge length of simulation domain "
+        "bounding box. This is useful when the simulation domain size is subwavelength.",
+        ge=1.0,
+    )
+
+    def _dl_max(self, sim_size: Tuple[float, 3]) -> float:
+        """Upper bound of grid size, constrained by `min_steps_per_sim_size`."""
+        return max(sim_size) / self.min_steps_per_sim_size
```

Comment:
To make this more general, would it be better to take the smaller value between max(sim_size) / self.min_steps_per_sim_size and lambda_min / min_steps_per_wvl?

Reply:
`wavelength / self.min_steps_per_wvl` can be different in different materials. Currently, the maximal step size is the minimal value of two.

Reply:
that part is implemented here: https://github.com/flexcompute/tidy3d/blob/dad7fc3e9eb1f0e06b1f8d81d77aee9d14cb06a7/tidy3d/components/grid/mesher.py#L549

Reply:
I see. That's what I was looking for. Thanks.